shepherd/shepherd.py

```diff
@@ -0,0 +1,438 @@
+from multiprocessing import Process
+import json
+import multiprocessing
+import os
+import signal
+import subprocess
+import sys
+import threading
+import time
+import yaml
+
+# --- config_loader.py ---
+
+
+def load_and_preprocess_config(filepath):
+    """Loads and preprocesses configuration from a YAML file."""
+    with open(filepath, 'r') as file:
+        config = yaml.safe_load(file)
+
+    preprocess_config(config, filepath)
+
+    print(f"DEBUG: Loaded and preprocessed config from {filepath}")
+    return config
+
+
+def preprocess_config(config, config_path):
+    """Automatically fills in missing stdout_path and stderr_path paths."""
+    services = config.get('services', {})
+    stdout_dir = config.get('output', {}).get('stdout_dir', '')
+    working_dir = os.path.dirname(os.path.abspath(config_path))
+
+    for service_name, details in services.items():
+        # Auto-fill log and error files if not specified
+        if 'stdout_path' not in details:
+            details['stdout_path'] = f"{service_name}_stdout.log"
+        if 'stderr_path' not in details:
+            details['stderr_path'] = f"{service_name}_stderr.log"
+
+        if stdout_dir:
+            details['stdout_path'] = os.path.join(stdout_dir, details['stdout_path'])
+            details['stderr_path'] = os.path.join(stdout_dir, details['stderr_path'])
+        else:
+            details['stdout_path'] = os.path.join(working_dir, details['stdout_path'])
+            details['stderr_path'] = os.path.join(working_dir, details['stderr_path'])
+
+        state_file_path = details.get('state', {}).get('file', {}).get('path', "")
+
+        if state_file_path:
+            details['state']['file']['path'] = os.path.join(working_dir, state_file_path)
+
+
+def validate_and_sort_programs(config):
+    print("DEBUG: Validating and sorting programs")
+    required_keys = ['services']
+
+    for key in required_keys:
```

Comment:
Any reason why are you checking here and not letting the exception be raised in whatever code uses it?
Also, if you are going to raise it, it is a `KeyError`.